File: lib/dotfiles/executor.rb
# frozen_string_literal: true

require 'English'
require 'fileutils'
require 'json'
require 'tmpdir'

module Dotfiles
    # Applies planned state changes and reports the resulting state of each one.
    class Executor
        WINDOWS_POWERSHELL = 'C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe'
        # Elevates a command through a single UAC prompt without elevating the
        # rest of the run. The command is handed over as a JSON file rather than
        # CLI arguments: PowerShell's parameter binder treats any argument
        # starting with "-" (e.g. "-NoProfile", "-Command") as a potential named
        # parameter, which breaks positional/remaining-argument binding for
        # commands that are themselves PowerShell invocations.
        # The heredoc delimiter is quoted to disable Ruby's escape processing:
        # an unquoted delimiter collapses '\"' to '"' before it ever reaches
        # PowerShell, silently breaking the embedded-quote escaping below.
        ELEVATOR_SCRIPT = <<~'POWERSHELL'
            param([Parameter(Mandatory = $true)][string]$PayloadPath)

            $payload = Get-Content -Raw -Path $PayloadPath | ConvertFrom-Json

            # Start-Process's -ArgumentList does not quote array elements that
            # contain spaces, so a multi-word argument (e.g. a "-Command" script)
            # gets split into separate tokens for the child process. Quoting each
            # element and joining into one string avoids that.
            $quotedArguments = @($payload.arguments) | ForEach-Object { '"' + ($_ -replace '"', '\"') + '"' }
            $startParameters = @{
                FilePath = $payload.file_path
                ArgumentList = ($quotedArguments -join ' ')
                Verb = 'RunAs'
                Wait = $true
                PassThru = $true
            }
            if ($payload.working_directory) {
                $startParameters.WorkingDirectory = $payload.working_directory
            }
            $process = Start-Process @startParameters
            exit $process.ExitCode
        POWERSHELL

        # @param repository_root [String] absolute repository path
        # @param home_directory [String] destination home directory
        # @param clean [Boolean] whether existing regular files may be removed
        # @param state_path [String] local command execution state path
        def initialize(repository_root:, home_directory: Dir.home, clean: false,
                       state_path: File.join(repository_root, '.local', 'state.json'))
            @repository_root = repository_root
            @home_directory = home_directory
            @clean = clean
            @state_store = StateStore.new(state_path)
            @elevator_script_path = File.join(Dir.tmpdir, 'dotfiles-elevate.ps1')
        end

        # Applies every selected state change in order.
        #
        # @param plan [Dotfiles::Plan]
        # @return [Array<Symbol>] results for each executed action
        def execute(plan)
            plan.actions.map { |action| execute_action(action) }
        end

        # Describes whether a state change is satisfied, pending, blocked, or unsupported.
        #
        # @param action [Dotfiles::Action]
        # @return [Symbol]
        def status(action)
            case action.name
            when :link_file
                link_status(action)
            when :copy_file
                copy_status(action)
            when :run_command
                command_status(action)
            else
                :unsupported
            end
        end

        private

        def execute_action(action)
            case action.name
            when :link_file
                link_file(action)
            when :copy_file
                copy_file(action)
            when :run_command
                run_command(action)
            else
                raise "Unsupported action: #{action.name}"
            end
        end

        def link_file(action)
            source, target = file_paths(action)

            raise "Source file does not exist: #{source}" unless File.file?(source)

            if File.symlink?(target)
                current_target = begin
                    File.realpath(target)
                rescue Errno::ENOENT
                    nil
                end
                return :already_linked if current_target == File.realpath(source)

                File.delete(target)
            elsif File.exist?(target)
                raise "Refusing to replace existing file: #{target}" unless @clean

                raise "Refusing to remove existing non-file path: #{target}" unless File.file?(target)

                File.delete(target)
            end

            FileUtils.mkdir_p(File.dirname(target))
            create_symlink(source, target, action.elevation)
            :linked
        end

        def create_symlink(source, target, elevation)
            return File.symlink(source, target) unless elevation == :admin

            # Windows only grants SeCreateSymbolicLinkPrivilege to standard users
            # when Developer Mode is on, so this machine needs elevation for it.
            #
            # The paths are embedded directly in the script text rather than
            # passed as trailing arguments: "-Command" does not bind trailing
            # CLI arguments to $args the way "-File" does, so they would
            # otherwise be silently ignored instead of reaching New-Item.
            script = "New-Item -ItemType SymbolicLink -Path '#{powershell_quote(target)}' " \
                     "-Target '#{powershell_quote(source)}' -Force | Out-Null"
            command = [WINDOWS_POWERSHELL, '-NoProfile', '-Command', script]
            return if run_elevated(command)

            raise "Elevated symlink creation failed: #{target}"
        end

        def powershell_quote(value)
            value.gsub("'", "''")
        end

        def link_status(action)
            source, target = file_paths(action)

            return :missing_source unless File.file?(source)
            return :pending unless File.exist?(target) || File.symlink?(target)
            return :blocked unless File.symlink?(target)

            current_target = begin
                File.realpath(target)
            rescue Errno::ENOENT
                nil
            end

            return :linked if current_target == File.realpath(source)

            :replaceable
        end

        def copy_file(action)
            source, target = file_paths(action)

            raise "Source file does not exist: #{source}" unless File.file?(source)
            raise "Refusing to replace existing non-file path: #{target}" if File.directory?(target)
            return :already_copied if File.file?(target) && !File.symlink?(target) && FileUtils.compare_file(source,
                                                                                                             target)

            FileUtils.mkdir_p(File.dirname(target))
            FileUtils.rm_f(target)
            FileUtils.cp(source, target)
            :copied
        end

        def copy_status(action)
            source, target = file_paths(action)

            return :missing_source unless File.file?(source)
            return :pending unless File.file?(target) && !File.symlink?(target)

            FileUtils.compare_file(source, target) ? :already_copied : :pending
        end

        def run_command(action)
            # Desired state is trusted repository code, so validation only protects the
            # command shape for now.
            command = action.parameters.fetch(:command)
            validate_command(command)
            fingerprint = action.fingerprint(repository_root: @repository_root)
            return :already_applied if @state_store.completed?(action, fingerprint)

            succeeded = if action.elevation == :admin
                            run_elevated(command)
                        else
                            system(*command, chdir: @repository_root)
                        end

            unless succeeded
                exit_code = $CHILD_STATUS.exitstatus || 'unknown'
                raise "Command failed with exit code #{exit_code}: #{command.join(' ')}"
            end

            @state_store.record(action, fingerprint)
            :executed
        end

        # Runs a command with a single UAC prompt, regardless of whether the
        # current process is already elevated. Start-Process auto-approves
        # elevation requests from an already-elevated parent, so this is safe
        # to call unconditionally rather than branching on current privilege.
        def run_elevated(command)
            resolved_command = command.each_with_index.map do |argument, index|
                if index.positive? && command[index - 1] == '-File'
                    File.expand_path(argument, @repository_root)
                else
                    argument
                end
            end
            payload_path = File.join(Dir.tmpdir, "dotfiles-elevate-#{$PROCESS_ID}.json")
            payload = {
                'file_path' => resolved_command.first,
                'arguments' => resolved_command.drop(1),
                'working_directory' => @repository_root
            }
            File.write(payload_path, JSON.generate(payload))
            File.write(@elevator_script_path, ELEVATOR_SCRIPT)
            system(WINDOWS_POWERSHELL, '-NoProfile', '-ExecutionPolicy',
                   'Bypass', '-File', @elevator_script_path, '-PayloadPath', payload_path)
        ensure
            FileUtils.rm_f(payload_path) if payload_path
        end

        def command_status(action)
            fingerprint = action.fingerprint(repository_root: @repository_root)
            @state_store.completed?(action, fingerprint) ? :already_applied : :planned
        end

        def validate_command(command)
            return if command.is_a?(Array) && !command.empty? && command.all?(String)

            raise ArgumentError, 'Command must be a non-empty array of strings.'
        end

        def file_paths(action)
            source = File.expand_path(action.parameters.fetch(:source), @repository_root)
            target = expand_target(action.parameters.fetch(:target))
            [source, target]
        end

        def expand_target(target)
            target = target.gsub(/%([^%]+)%/) do
                variable = Regexp.last_match(1)
                ENV.fetch(variable) { raise "Environment variable is not set: #{variable}" }
            end

            relative_target = target.sub(%r{\A~[\\/]}, '')
            return File.join(@home_directory, relative_target) if relative_target != target

            File.expand_path(target)
        end
    end
end

File: src/dotfiles/executor.py
"""Apply planned filesystem and command actions."""

from __future__ import annotations

import filecmp
import json
import os
import re
import shutil
import sys
import tempfile
from pathlib import Path
from typing import Any
from uuid import uuid4

from dotfiles.action import Action
from dotfiles.command_runner import CommandRunner
from dotfiles.context import Context
from dotfiles.plan import Plan
from dotfiles.state_store import StateStore


class Executor:
    """Apply actions while protecting unmanaged files by default."""

    WINDOWS_POWERSHELL = r"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe"
    ELEVATOR_SCRIPT = r"""
param([Parameter(Mandatory = $true)][string]$PayloadPath)

$ErrorActionPreference = 'Stop'
$payload = Get-Content -Raw -Path $PayloadPath | ConvertFrom-Json
$quotedArguments = @($payload.arguments) | ForEach-Object { '"' + ($_ -replace '"', '\"') + '"' }
$startParameters = @{
    FilePath = $payload.file_path
    ArgumentList = ($quotedArguments -join ' ')
    Verb = 'RunAs'
    Wait = $true
    PassThru = $true
}
if ($payload.working_directory) {
    $startParameters.WorkingDirectory = $payload.working_directory
}
try {
    $process = Start-Process @startParameters
    if ($null -eq $process) {
        throw 'Start-Process did not return a process.'
    }
    exit [int]$process.ExitCode
} catch {
    Write-Error $_
    exit 1
}
""".lstrip()

    def __init__(
        self,
        repository_root: Path,
        home_directory: Path | None = None,
        clean: bool = False,
        state_path: Path | None = None,
        runner: CommandRunner | Any | None = None,
        platform_name: str | None = None,
    ) -> None:
        self.repository_root = Path(repository_root)
        self.home_directory = Path.home() if home_directory is None else Path(home_directory)
        self.clean = clean
        self.platform_name = platform_name or Context.determine_platform(sys.platform)

        self.runner = runner or CommandRunner()
        self.state_store = StateStore(
            Path(state_path) if state_path is not None else self.repository_root / ".local/state.json"
        )

    def execute(self, plan: Plan) -> list[str]:
        """Apply every action in order and return its result status."""
        return [self._execute_action(action) for action in plan.actions]

    def status(self, action: Action) -> str:
        """Describe whether an action is satisfied, pending, or blocked."""
        handlers = {
            "link_file": self._link_status,
            "copy_file": self._copy_status,
            "run_command": self._command_status,
        }

        handler = handlers.get(action.name)

        return handler(action) if handler else "unsupported"

    def _execute_action(self, action: Action) -> str:
        handlers = {
            "link_file": self._link_file,
            "copy_file": self._copy_file,
            "run_command": self._run_command,
        }

        handler = handlers.get(action.name)
        if handler is None:
            raise ValueError(f"Unsupported action: {action.name}")

        return handler(action)

    def _link_file(self, action: Action) -> str:
        source, target = self._file_paths(action)

        if not source.is_file():
            raise FileNotFoundError(f"Source file does not exist: {source}")

        if target.is_symlink():
            current_target = self._resolved_link(target)

            if current_target == source.resolve():
                return "already_linked"

            if not self.clean:
                raise FileExistsError(f"Refusing to replace existing symlink: {target}")
        elif target.exists():
            if not self.clean:
                raise FileExistsError(f"Refusing to replace existing file: {target}")
            if not target.is_file():
                raise IsADirectoryError(f"Refusing to remove existing non-file path: {target}")

        target.parent.mkdir(parents=True, exist_ok=True)
        self._replace_with_symlink(source, target, action.elevation)

        return "linked"

    def _replace_with_symlink(self, source: Path, target: Path, elevation: str) -> None:
        staged_target = target.with_name(f".{target.name}.dotfiles-{uuid4().hex}")

        try:
            self._create_symlink(source, staged_target, elevation)
            os.replace(staged_target, target)
        finally:
            staged_target.unlink(missing_ok=True)

    def _create_symlink(self, source: Path, target: Path, elevation: str) -> None:
        try:
            target.symlink_to(source)
            return
        except PermissionError:
            if elevation != "admin":
                raise

        if self.platform_name == "windows":
            script = (
                f"New-Item -ItemType SymbolicLink -Path '{self._powershell_quote(target)}' "
                f"-Target '{self._powershell_quote(source)}' -Force | Out-Null"
            )
            command = [self.WINDOWS_POWERSHELL, "-NoProfile", "-Command", script]
        else:
            command = ["ln", "-s", "--", str(source), str(target)]

        self._run_elevated(command)

        if self._resolved_link(target) != source.resolve():
            raise RuntimeError(f"Elevated symlink creation did not create the requested link: {target}")

    @staticmethod
    def _powershell_quote(value: Path) -> str:
        return str(value).replace("'", "''")

    def _link_status(self, action: Action) -> str:
        source, target = self._file_paths(action)

        if not source.is_file():
            return "missing_source"
        if not target.exists() and not target.is_symlink():
            return "pending"
        if not target.is_symlink():
            return "blocked"

        return "linked" if self._resolved_link(target) == source.resolve() else "replaceable"

    def _copy_file(self, action: Action) -> str:
        source, target = self._file_paths(action)

        if not source.is_file():
            raise FileNotFoundError(f"Source file does not exist: {source}")
        if target.is_dir() and not target.is_symlink():
            raise IsADirectoryError(f"Refusing to replace existing non-file path: {target}")

        files_match = target.is_file() and not target.is_symlink() and filecmp.cmp(source, target, shallow=False)
        if files_match:
            return "already_copied"

        if (target.exists() or target.is_symlink()) and not self.clean:
            raise FileExistsError(f"Refusing to replace existing file: {target}")

        target.parent.mkdir(parents=True, exist_ok=True)
        self._replace_with_copy(source, target)

        return "copied"

    @staticmethod
    def _replace_with_copy(source: Path, target: Path) -> None:
        file_descriptor, temporary_name = tempfile.mkstemp(
            prefix=f".{target.name}.dotfiles-",
            dir=target.parent,
        )
        os.close(file_descriptor)
        staged_target = Path(temporary_name)

        try:
            shutil.copy2(source, staged_target)
            os.replace(staged_target, target)
        finally:
            staged_target.unlink(missing_ok=True)

    def _copy_status(self, action: Action) -> str:
        source, target = self._file_paths(action)

        if not source.is_file():
            return "missing_source"
        if not target.is_file() or target.is_symlink():
            return "pending"

        files_match = filecmp.cmp(source, target, shallow=False)

        return "already_copied" if files_match else "pending"

    def _run_command(self, action: Action) -> str:
        command = action.parameters.get("command")
        self._validate_command(command)

        fingerprint = action.fingerprint(self.repository_root)
        if self.state_store.completed(action, fingerprint):
            return "already_applied"

        if action.elevation == "admin":
            self._run_elevated(command)
        else:
            self.runner.interactive(command, cwd=self.repository_root)

        self.state_store.record(action, fingerprint)

        return "executed"

    def _run_elevated(self, command: list[str]) -> None:
        if self.platform_name == "linux":
            self.runner.interactive(["sudo", "--", *command], cwd=self.repository_root)
            return

        if self.platform_name != "windows":
            raise RuntimeError(f"Administrator elevation is not supported on {self.platform_name}.")

        self._run_windows_elevated(command)

    def _run_windows_elevated(self, command: list[str]) -> None:
        resolved_command = list(command)

        for index, argument in enumerate(resolved_command):
            if index and resolved_command[index - 1] == "-File":
                resolved_command[index] = str(self.repository_root / argument)

        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", suffix=".json", prefix="dotfiles-elevate-", delete=False
        ) as payload_file:
            payload_path = Path(payload_file.name)
            payload = {
                "file_path": resolved_command[0],
                "arguments": resolved_command[1:],
                "working_directory": str(self.repository_root),
            }

            json.dump(payload, payload_file)

        script_path = Path(tempfile.gettempdir()) / f"dotfiles-elevate-{os.getpid()}.ps1"

        try:
            script_path.write_text(self.ELEVATOR_SCRIPT, encoding="utf-8")
            self.runner.interactive(
                [
                    self.WINDOWS_POWERSHELL,
                    "-NoProfile",
                    "-ExecutionPolicy",
                    "Bypass",
                    "-File",
                    str(script_path),
                    "-PayloadPath",
                    str(payload_path),
                ]
            )
        finally:
            payload_path.unlink(missing_ok=True)
            script_path.unlink(missing_ok=True)

    def _command_status(self, action: Action) -> str:
        fingerprint = action.fingerprint(self.repository_root)

        return "already_applied" if self.state_store.completed(action, fingerprint) else "planned"

    @staticmethod
    def _validate_command(command: Any) -> None:
        valid = isinstance(command, list) and command and all(isinstance(item, str) for item in command)
        if not valid:
            raise ValueError("Command must be a non-empty array of strings.")

    def _file_paths(self, action: Action) -> tuple[Path, Path]:
        source = self.repository_root / action.parameters["source"]
        target = self._expand_target(action.parameters["target"])

        return source, target

    def _expand_target(self, target: str) -> Path:
        def replace_environment_variable(match: re.Match[str]) -> str:
            name = match.group(1)
            value = os.environ.get(name)

            if value is None:
                raise RuntimeError(f"Environment variable is not set: {name}")

            return value

        expanded = re.sub(r"%([^%]+)%", replace_environment_variable, target)

        home_relative = re.match(r"^[~][\\/]", expanded)
        if home_relative:
            return self.home_directory / expanded[2:]

        return Path(expanded).expanduser().absolute()

    @staticmethod
    def _resolved_link(path: Path) -> Path | None:
        try:
            return path.resolve(strict=True)
        except FileNotFoundError:
            return None
